httpclient5/src/main/java/org/apache/hc/client5/http/impl/routing/FailoverProxySelector.java

```diff
@@ -0,0 +1,127 @@
+ * calls to select(). If a connection fails, connectFailed() is called on the
+ * active ProxySelector.
+ * <p>
+ * This implementation is useful in environments where multiple proxy servers may be
+ * used, and where it is desirable to failover to a different proxy if the current
+ * one fails.
+ *
+ * @since 5.3
+ */
+public class FailoverProxySelector extends ProxySelector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FailoverProxySelector.class);
+
+    private final List<ProxySelector> selectors;
+    private final AtomicInteger currentIndex;
+
+
+    /**
+     * Constructs a FailoverProxySelector from the given list of ProxySelectors.
+     *
+     * @param selectors the list of ProxySelectors to use.
+     * @throws IllegalArgumentException if the list is null or empty.
+     */
+    public FailoverProxySelector(final List<ProxySelector> selectors) {
+        if (selectors == null || selectors.isEmpty()) {
+            throw new IllegalArgumentException("At least one ProxySelector is required");
+        }
+        this.selectors = new ArrayList<>(selectors);
+        currentIndex = new AtomicInteger(0);
+
+    }
+
+    /**
+     * Selects a list of proxies for the given URI by delegating to the next
+     * available ProxySelector in the list, until a non-empty proxy list is
+     * returned. Once a non-empty proxy list is returned, it becomes the active
+     * ProxySelector and is used for subsequent calls to select().
+     *
+     * @param uri the URI to select a proxy for.
+     * @return a list of proxies for the given URI.
+     */
+    @Override
+    public List<Proxy> select(final URI uri) {
+        List<Proxy> result = Collections.emptyList();
+        for (int i = 0; i < selectors.size(); i++) {
+            final ProxySelector selector = selectors.get((i + currentIndex.get()) % selectors.size());
+            try {
+                result = selector.select(uri);
+                if (!result.isEmpty()) {
+                    currentIndex.set((i + currentIndex.get()) % selectors.size());
+                    break;
+                }
+            } catch (final Exception e) {
+                // ignore and try the next selector
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Exception caught while selecting proxy for URI {}: {}", uri, e.getMessage());
+
+                }
+            }
+        }
+        return result;
+    }
+
+
+    /**
+     * Notifies the active ProxySelector that a connection to the given URI
+     * using the given proxy has failed.
+     *
+     * @param uri the URI that failed to connect.
+     * @param sa  the SocketAddress of the proxy that failed to connect.
+     * @param ioe the IOException that resulted from the failed connection.
+     */
+    @Override
+    public void connectFailed(final URI uri, final SocketAddress sa, final IOException ioe) {
+        final ProxySelector selector = selectors.get(currentIndex.get());
```

Comment:
@arturobernalg This approach can produce unexpected results in the selector is being used concurrently by multiple threads. The current index can easily change from another thread while the `#select` method is still being called. As much as I hate `ThreadLocal`s I suppose this implementation must maintain a separate state per each thread of execution.

Reply:
@arturobernalg Better, but we can make it better still. 
1. Use `ThreadLocal` to store the currently chosen selector
2. Unset the `ThreadLocal` once the `#select` call has been completed successfully or unsuccessfully. 
3. Use `AtomicInteger` to store the shared current index